File: core0/us/sdk/wallet/cs/us/gov/socket/datagram.cs
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Text;

namespace us.gov.socket
{
    public class datagram
    {
        static readonly int offset_channel = 0;
        static readonly int size_channel = 2;
        static readonly int offset_size = offset_channel + size_channel;
        static readonly int size_size = 4;
        static readonly int offset_service = offset_size + size_size;
        static readonly int size_service = 2;
        static readonly int h = size_channel + size_size + size_service;
        static readonly int maxsize = 20000000;
        public static short system_channel = 0;

        byte[] bytes;
        public int dend;
        public short service;

        public datagram()
        {
            bytes = new byte[h];
            dend = 0;
        }

        public datagram(UInt16 service)
        {
            bytes = new byte[h];
            encode_channel();
            encode_size(h);
            encode_service(service);
            dend = bytes.Length;
        }

        public datagram(UInt16 service, String msg)
        {
            int size = h + msg.Length;
            bytes = new byte[size];
            encode_channel();
            encode_size(size);
            encode_service(service);
            //System.arraycopy(msg.getBytes(), 0, bytes, h, msg.Length);
            Buffer.BlockCopy(Encoding.ASCII.GetBytes(msg), 0, bytes, h, msg.Length);
            dend = bytes.Length;
        }

        public datagram(UInt16 service, short payload)
        {
            int size = h + 2;
            bytes = new byte[size];
            encode_channel();
            encode_size(size);
            encode_service(service);
            bytes[h] = (byte)(payload & 0xff);
            bytes[h + 1] = (byte)(payload >> 8);
            dend = bytes.Length;    
        }

        /*
        public static byte[] hexStringToByteArray(String s)
        {
            int len = s.Length;
            byte[] data = new byte[len / 2];
            for (int i = 0; i < len; i += 2)
            {
                data[i / 2] = (byte)((Char.digit(s[i], 16) << 4) + Char.digit(s[i + 1], 16));
            }
            return data;
        }
        */

        public datagram(UInt16 service, byte[] addr, byte[] content)
        {
            if (addr.Length != 20) return;
            int payload_sz = 20 + content.Length;
            int size = h + payload_sz;
            bytes = new byte[size];
            encode_channel();
            encode_size(size);
            encode_service(service);     
            Buffer.BlockCopy(addr, 0, bytes, h, 20);            
            Buffer.BlockCopy(content, 0, bytes, h + 20, content.Length);
            dend = bytes.Length;
        }

        public datagram(UInt16 service, byte[] addr)
        {
            if (addr.Length != 20) return;
            int payload_sz = 20;
            int size = h + payload_sz;
            bytes = new byte[size];
            encode_channel();
            encode_size(size);
            encode_service(service);
            //System.arraycopy(addr, 0, bytes, h, 20);
            Buffer.BlockCopy(addr, 0, bytes, h, 20);            
            dend = bytes.Length;
        }

        public datagram(UInt16 service, byte[] addr, String s1)
        {
            if (addr.Length != 20) return;
            int payload_sz = 20 + s1.Length;
            int size = h + payload_sz;
            bytes = new byte[size];
            encode_channel();
            encode_size(size);
            encode_service(service);
            Buffer.BlockCopy(addr, 0, bytes, h, 20);            
            Buffer.BlockCopy(Encoding.ASCII.GetBytes(s1), 0, bytes, h + 20, s1.Length);
            dend = bytes.Length;
        }

        public void encode_channel()
        {
            Debug.Assert(bytes.Length >= offset_channel + 2);
            byte[] t = BitConverter.GetBytes(system_channel);
            bytes[offset_channel] = t[0];
            bytes[offset_channel + 1] = t[1];
        }

        public short decode_channel()
        {
            Debug.Assert(bytes.Length > 2);
            byte[] t = new byte[2];
            t[0] = bytes[offset_channel];
            t[1] = bytes[offset_channel + 1];
            /*
            MemoryStream bb = new MemoryStream();
            using (BinaryWriter writer = new BinaryWriter(bb))
            {
                writer.Write(t);
            }
            */
            return BitConverter.ToInt16(t, 0);
        }

        public void encode_size(int sz)
        {
            Debug.Assert(bytes.Length >= offset_size + 3);
            bytes[offset_size] = (byte)(sz & 0xff);
            bytes[offset_size + 1] = (byte)(sz >> 8 & 0xff);
            bytes[offset_size + 2] = (byte)(sz >> 16 & 0xff);
            bytes[offset_size + 3] = (byte)(sz >> 24 & 0xff);
        }

        public UInt32 decode_size()
        {
            Debug.Assert(bytes.Length > 3);
            UInt32 sz = bytes[offset_size];
            sz |= (uint)(bytes[offset_size + 1] << 8);
            sz |= (uint)(bytes[offset_size + 2] << 16);
            sz |= (uint)(bytes[offset_size + 3] << 24);
            return sz;
        }

        public long size()
        {
            return bytes.Length;
        }

        public void encode_service(UInt16 svc)
        {
            Debug.Assert(bytes.Length >= offset_service + 2);
            bytes[offset_service] = (byte)(svc & 0xff);
            bytes[offset_service + 1] = (byte)(svc >> 8 & 0xff);
        }

        public UInt16 decode_service()
        {
            Debug.Assert(bytes.Length >= h);
            UInt16 svc = bytes[offset_service];
            svc |= (ushort)(bytes[offset_service + 1] << 8);
            return svc;
        }

        public bool completed()
        {
            return dend == bytes.Length && bytes.Length > 0;
        }

        public String sendto(Socket sock)
        {
            if (bytes.Length >= maxsize)
            {
                return "KO - Datagram is too big.";
            }
            try
            {                
                sock.Send(bytes);
                //int bytesRec = sock.Receive(bytes);
                //return Encoding.UTF8.GetString(bytes, 0, bytes.Length);
            }
            catch (Exception e)
            {
                return e.Message;
            }
            return null;
        }

        public static int timeout = 3000;

        public String recvfrom(Socket s)
        {
            
            //System.out.println("dgram::recvfrom dend="+dend);
            try
            {                
                //DataInputStream input = new DataInputStream(s.getInputStream());
                if (dend < h)
                {
                    //int nread = input.read(bytes, dend, h - dend);
                    int nread = s.Receive(bytes, dend, h - dend, SocketFlags.None);
                    //System.out.println("dgram::recvfrom nread="+nread);
                    if (nread <= 0)
                    {
                        return null;
                    }
                    dend += nread;
                    if (dend < h) return null;
                    short ch = decode_channel();
                    if (ch != system_channel)
                    {
                        return "evil. channel mismatch " + ch + "!=" + system_channel + "(system channel)";
                    }
                    int sz = (int)decode_size();
                    if (sz > maxsize)
                    {
                        return "evil. too big";
                    }
                    if (sz < h)
                    {
                        return "evil. too small";
                    }
                    byte[] m = new byte[sz];
                    Buffer.BlockCopy(bytes, 0, m, 0, h);
                    bytes = m;
                    service = (short)decode_service();
                    if (dend == sz)
                    {
                        return null;
                    }
                }
                //int nread = input.read(bytes, dend, bytes.Length - dend);
                int nread2 = s.Receive(bytes, dend, bytes.Length - dend, SocketFlags.None);
                if (nread2 <= 0)
                {
                    return "KO - Timeout waiting for data from peer. or Incoming datagram is too big";
                }
                dend += nread2;
                return null;
            }
            catch (Exception e)
            {
                return "KO - " + e.Message;
            }
        }

        public String parse_string()
        {
            byte[] b = new byte[dend - h];
            //System.arraycopy(bytes, h, b, 0, dend - h);
            Buffer.BlockCopy(bytes, h, b, 0, dend - h);
            return Encoding.UTF8.GetString(b, 0, b.Length); ;
        }

        /*
        public short parse_uint16()
        {
            if (bytes.Length != h + 2) return 0;
            byte[] t = new byte[2];
            t[0] = bytes[h];
            t[1] = bytes[h + 1];
            ByteBuffer bb = ByteBuffer.wrap(t);
            bb.order(ByteOrder.LITTLE_ENDIAN);
            return bb.getShort();
        }
        */
        public byte[] parse_bin()
        {
            byte[] t = new byte[dend - h];
            Buffer.BlockCopy(bytes, h, t, 0, dend - h);
            return t;
        }        

        static char[] x = { 'P', 'E', 'Q', '_' };

        public String service_str()
        {
            return service_str(service);
        }

        public String service_str(short svc)
        {
            String os = null;
            if (svc >= 100)
            {
                os = x[svc & 3] + ".";
                int m = svc >> 2;
                int l = m / 100;
                os += l + "." + m % 100;
            }
            else
            {
                os = "wallet";
            }
            return os;
        }
    }
}

File: core0/us/sdk/wallet/java/us/gov/socket/datagram.java
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.gov.socket;
import java.util.ArrayList;                                                                    // ArrayList
import org.bouncycastle.util.Arrays;                                                           // Arrays
import us.gov.crypto.base58;                                                                   // base58
import us.gov.io.blob_writer_t;                                                                // blob_writer_t
import java.io.BufferedReader;                                                                 // BufferedReader
import java.nio.ByteBuffer;                                                                    // ByteBuffer
import java.nio.ByteOrder;                                                                     // ByteOrder
import us.CFG;                                                                                 // CFG
import java.io.DataInputStream;                                                                // DataInputStream
import java.io.DataOutputStream;                                                               // DataOutputStream
import java.security.GeneralSecurityException;                                                 // GeneralSecurityException
import us.gov.crypto.ripemd160.hash_t;                                                         // hash_t
import static us.gov.socket.types.*;                                                           // *
import us.stdint.*;                                                                            // *
import us.tuple.*;                                                                             // *
import java.io.InputStream;                                                                    // InputStream
import java.io.InputStreamReader;                                                              // InputStreamReader
import java.io.IOException;                                                                    // IOException
import us.ko;                                                                                  // ko
import static us.ko.ok;                                                                        // ok
import java.io.OutputStream;                                                                   // OutputStream
import us.pair;                                                                                // pair
import java.io.PrintWriter;                                                                    // PrintWriter
import java.security.PrivateKey;                                                               // PrivateKey
import java.security.PublicKey;                                                                // PublicKey
import us.gov.crypto.ripemd160;                                                                // ripemd160
import us.gov.crypto.sha256;                                                                   // sha256
import java.net.Socket;                                                                        // Socket
import us.gov.crypto.symmetric_encryption;                                                     // symmetric_encryption

public class datagram {
    static final int offset_channel = 0;
    static final int size_channel = 2;
    static final int offset_size = offset_channel+size_channel;
    static final int size_size = 4;
    static final int offset_service = offset_size+size_size;
    static final int size_service = 2;
    static final int offset_sequence = offset_service+size_service;
    static final int size_sequence = 2;

    public static final int h = size_channel + size_size + size_service + size_sequence;
    public static final int maxsize = CFG.MAX_SIZE_DATAGRAM;

    public static final ko KO_0001 = new ko("KO 0001 channel mismatch.");
    public static final ko KO_0002 = new ko("KO 0002 too big.");
    public static final ko KO_0003 = new ko("KO 0003 too small.");
    public static final ko KO_4038 = new ko("KO 5933 invalid size");
    public static final ko KO_4921 = new ko("KO 4921 ::recv<0");
    public static final ko KO_9021 = new ko("KO 9021 ::recv 0");
    public static final ko KO_20292 = new ko("KO 20292 Size would buffer overflow.");

    public static interface dispatcher_t {
        boolean dispatch(datagram d); //return true is datagram has been processed
    }

    static void log(final String line) {                            //--strip
        CFG.log_gov_socket("datagram: " + line);                    //--strip
    }                                                               //--strip

    public datagram(final channel_t channel, int mode) {
        switch(mode) {
            case 0: {  //used in encryption
                bytes = new byte[h];
                encode_channel(channel);
                //encode_size(sz);
                service = new svc_t(0);
                encode_service(service);
                encode_sequence(new seq_t(0));
            }
            break;
            case 1: { //used in decryption
            }
            break;
            case 2: { //used in recv
                bytes = new byte[h];
                dend = 0;
            }
            break;
        }
    }


    public datagram(final channel_t channel_, final svc_t service_, final seq_t sequence, int payload_sz) {
        assert channel_ != null;
        assert service_ != null;
        assert sequence != null;
        bytes = new byte[h + payload_sz];
        service = service_;
        encode_channel(channel_);
        encode_size(h + payload_sz);
        encode_service(service);
        encode_sequence(sequence);
        dend = bytes.length;
        log("new datagram svc" + service.value + " seq " + sequence.value + " sz " + bytes.length + " payload sz " + payload_sz + " ch " + decode_channel().value); //--strip
    }

    public pair<ko, datagram> encrypt(symmetric_encryption se) {
        log("encrypt"); //--strip
        datagram d = new datagram(decode_channel(), 0);
        pair<ko, byte[]> r = se.encrypt(bytes);
        if (ko.is_ko(r.first)) {
            return new pair<ko, datagram>(r.first, null);
        }
        d.bytes = Arrays.concatenate(d.bytes, r.second);
        d.dend = d.bytes.length;
        d.encode_size(d.dend);
        return new pair<ko, datagram>(ok, d);
    }

    public pair<ko, datagram> decrypt(channel_t channel, symmetric_encryption se) {
        if (service.value != 0) {
            ko r = new ko("KO 81920 svc should be 0 for encrypted dgram.");
            log("decrypt dgram " + r.msg + " " + service.value); //--strip
            return new pair<ko, datagram>(r, null);
        }
        log("decrypt dgram " + bytes.length); //--strip
        if (bytes.length < h) {
            ko r = new ko("KO 58948 Invalid dgram.");
            log("decrypt dgram " + r.msg + " " + service.value); //--strip
            return new pair<ko, datagram>(r, null);
        }
        datagram d = new datagram(channel, 1);
        {
            pair<ko, byte[]> r = se.decrypt(bytes, h, bytes.length - h);
            if (ko.is_ko(r.first)) {
                return new pair<ko, datagram>(r.first, null);
            }
            d.bytes = r.second;
        }
        d.dend = d.bytes.length;
        if (d.bytes.length < h) {
            ko r = new ko("KO 25994 Invalid decrypted size.");
            log(r.msg); //--strip
            return new pair<ko, datagram>(r, null);
        }
        channel_t ch = d.decode_channel();
        if (ch.value != channel.value) {
            if (channel.value != all_channels.value) { //accepting any channel
                log(KO_0001.msg + " evil1 enc dgram_ch:" + ch.value + " sys:" + channel.value); //--strip
                return new pair<ko, datagram>(KO_0001, null);
            }
            log("enc dgram arrived targeting ALL channels."); //--strip
        }
        else {
            log("enc dgram arrived targeting my channel. " + ch.value);  //--strip
        }

        int sz = d.decode_size();
        if (sz != d.bytes.length) {
            log(KO_4038.msg + "DGRAM DECRYPT KO 6"); //--strip
            return new pair<ko, datagram>(KO_4038, null);
        }
        if (sz < h) {
            log(KO_4038.msg + "DGRAM DECRYPT KO 7"); //--strip
            return new pair<ko, datagram>(KO_4038, null);
        }
        d.service = d.decode_service();
        return new pair<ko, datagram>(ok, d);
    }

    public boolean validate_payload(byte[] payload) {
        return h + blob_writer_t.blob_size(payload) < maxsize;
    }

    public boolean in_progress() { return dend > 0 && dend < size(); }

    public void encode_channel(channel_t channel) {
        assert bytes.length >= offset_channel + 2;
        log("writing dgram channel " + channel.value); //--strip
        channel.write(bytes, offset_channel);
    }

    public channel_t decode_channel() {
        assert(bytes.length >= 2);
        channel_t o = new channel_t();
        o.read(bytes, offset_channel);
        return o;
    }

    public void encode_size(int sz) {
        log("encoding size=" + sz); //--strip
        assert(bytes.length > offset_size + 4);
        new uint32_t(sz).write(bytes, offset_size);
    }

    public int decode_size() {
        assert bytes.length >= 4;
        return (int)uint32_t.from(bytes, offset_size).value;
    }

    public int size() {
        return bytes.length;
    }

    public void encode_service(svc_t svc) {
        assert bytes.length >= offset_service + 2;
        svc.write(bytes, offset_service);
    }

    public svc_t decode_service() {
        assert bytes.length >= h;
        svc_t svc = new svc_t();
        svc.read(bytes, offset_service);
        return svc;
    }

    public void encode_sequence(seq_t seq) {
        assert(bytes.length >= offset_sequence + 2);
        seq.write(bytes, offset_sequence);
    }

    public seq_t decode_sequence() {
        assert bytes.length >= h;
        seq_t seq = new seq_t();
        seq.read(bytes, offset_sequence);
        return seq;
    }

    public boolean completed() {
        return dend == bytes.length && bytes.length > 0;
    }

    public static int timeout = 3000;

    ko recvfrom(Socket s, channel_t channel, busyled_t busyled) {
        log("recvfrom. ch " + channel.value); //--strip
        log("recvfrom timeout is " + timeout); //--strip
        try {
            log("recvfrom socket t/o: " + s.getSoTimeout()); //--strip
            if (dend < h) {
                log("READ-A"); //--strip
                busyled.reset();
                int nread = s.getInputStream().read(bytes, dend, h - dend);
                busyled.set();
                log("/READ-A"); //--strip
                if (nread < 0) {
                    log(KO_4921.msg+" nread < 0; (" + nread + ")."); //--strip
                    return KO_4921;
                }
                dend += nread;
                if (dend < h) {
                    log("dend is < h " + dend); //--strip
                    return ok;
                }
                channel_t ch = decode_channel();
                if (ch.value != channel.value) {
                    if (channel.value != all_channels.value) { //accepting any channel
                        log(KO_0001.msg + " evil1 dgram_ch:" + ch.value + " sys:" + channel.value); //--strip
                        return KO_0001; //"evil. channel mismatch " + ch + "!=" + system_channel + "(system channel)";
                    }
                    log("enc dgram arrived targeting ALL channels."); //--strip
                }
                else {
                    log("enc dgram arrived targeting my channel. " + ch.value);  //--strip
                }
                int sz = decode_size();
                if (sz > maxsize) {
                    log("evil2"); //--strip
                    return KO_0002; //"evil. too big";
                }
                if (sz < h) {
                    log("evil3"); //--strip
                    return KO_0003; //"evil. too small";
                }
                byte[] m = new byte[sz];
                System.arraycopy(bytes, 0, m, 0, h);
                bytes = m;
                service = decode_service();
                if (dend == sz) {
                    log("return Completed OK dend=" + dend); //--strip
                    return ok;
                }
                log("Received header for dgram sz=" + sz + " svc=" + service.value);  //--strip
            }
            log("READ-B " + (bytes.length - dend)); //--strip
            busyled.reset();
            int nread = s.getInputStream().read(bytes, dend, bytes.length - dend);  //-1 if there is no more data because the end of the stream has been reached. https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read()
            busyled.set();
            log("/READ-B"); //--strip
            if (nread < 0) {
                log(KO_4921.msg + " nread < 0; (" + nread + ")."); //--strip
                return KO_4921;
            }
            dend += nread;
            log("return OK read chunk " + nread + " dend=" + dend); //--strip
            return ok;
        }
        catch(java.net.SocketTimeoutException e) {
            log("KO 6912 - Timeout Exception " + e.getMessage()); //--strip
            return new ko("KO 6912 " + e.getMessage());
        }
        catch(IOException e) {
            log("KO 6251 - IOException " + e.getMessage()); //--strip
            return new ko("KO 6251 " + e.getMessage());
        }
        catch(Exception e) {
            log("KO 6651 - Exception " + e.getMessage()); //--strip
            return new ko("KO 6651 " + e.getMessage());
        }
    }

    public hash_t compute_hash() {
        return ripemd160.hash(bytes);
    }

    public hash_t compute_hash_payload() {
        byte[] ed = java.util.Arrays.copyOfRange(bytes, h, bytes.length-h);
        return ripemd160.hash(ed);
    }

    public byte[] bytes;
    //public static channel_t system_channel = CFG.CHANNEL;
    public static channel_t all_channels = new channel_t(65535);
    public svc_t service;
    public int dend;
}

